### PR TITLE
TUBii: First commit of a TUBii version of PGT.

### PR DESCRIPTION
**Pull request by icoulter**
First commit of a TUBii version of PGT. Appears to work as expected.
Needs an accompanying update to TUBii's firmware.

### CODE REVIEW

**Comment by tlatorre-uchicago** [comment]
Lemme know when the firmware update is done, and I'll merge.

**Comment by icoulter** [comment]
The testing of this didn't work due to a couple of issues (some related to this PR: I got two of the trigger bits swapped by mistake; some environmental: the firmware was expecting Ed's keep-alive pulse which hasn't been merged yet). We're going to test it again on Thursday with these issues fixed when the detector is at low voltage.

**Comment by ecaden** [comment]
I'm not sure yet that we're ramping down on Thursday.

--
Erica Caden, Ph.D.
SNOLAB Research Scientist
SNO+ Detector Manager
ecaden@snolab.ca
+1.705.692.7000 x2253

**Comment by icoulter** [comment]
Ah, your email on Friday said that there would be time available.

**Comment by tlatorre-uchicago** [comment]
OK. I'll hold off until then. Why does the PGT depend on the keep alive?

**Comment by icoulter** [comment]
It doesn't. I was rather silly and added the commits to the firmware in a linear way. I only realised after testing otherwise I could have worked around it.

**Comment by icoulter** [comment]
Got the chance to test it now as we're on LV due to the lightning storm. There's one minor bug fix in the firmware that I half expected: the internal clock is at the wrong rate and can't be changed in the compiler I ended up sticking with which means the rate was slower than expected by a factor of 2. This is easy to fix in the firmware though. The ORCA PR is fine.

File: Source/Objects/Custom Hardware/SNO+/TUBii/TUBiiModel.m [hunk 1]
```diff
@@ -189,6 +189,10 @@ - (void) setPrescaleTrigger_Mask:(uint32_t)mask ByFactor:(uint32_t)factor {
     NSString* const command = [NSString stringWithFormat:@"SetPrescaleTrigger %u %u",factor,mask];
     [self sendOkCmd:command];
 }
+- (void) setTUBiiPGT_Rate:(float)rate {
+    NSString* const command = [NSString stringWithFormat:@"SetTUBiiPGT %f 999999999",rate];
```
Comment (tlatorre-uchicago) [inline Source/Objects/Custom Hardware/SNO+/TUBii/TUBiiModel.m]: Just curious, what is the 999999999 for?

**Comment by tlatorre-uchicago** [comment]
I think it would be nice to have the ability to set the pulser rate independently of enabling it. How does that work currently? The way the MTCD currently works you can set a rate for the pulser, and then enable/disable the pulser.

File: Source/Objects/Custom Hardware/SNO+/TUBii/TUBiiModel.m (hunk 1 above)
Comment (icoulter) [inline Source/Objects/Custom Hardware/SNO+/TUBii/TUBiiModel.m]: Ah, forgot about this. The pulser needs a finite number of pulses to send. I'd hardcoded it to a large number.

**Comment by icoulter** [comment]
Currently you choose a rate and then load the settings. To disable it you'd need to set the rate to zero and load or else just disable it from the trigger mask.

**Comment by icoulter** [comment]
This should hopefully address your comments. I've changed the Send To TUBii button into two separate Start and Stop buttons.
I've also removed the hard coded number of pulses from ORCA into the firmware itself.

**Comment by tlatorre-uchicago** [comment]
Please squash the commit into a single one with a commit message like:

```
SNO+: [description]

[longer description]
```

and I'll merge.

**Comment by icoulter** [comment]
What's the command?

Quoting tlatorre-uchicago <notifications@github.com>:

**Comment by tlatorre-uchicago** [comment]
The way I usually do it (I'm sure there is an easier way):

```
$ git checkout master
$ git checkout -b new_branch
$ git merge --squash TUBiiPGT
$ git commit
$ git push -f origin new_branch:TUBiiPGT
```

Might want to make sure you copy the whole directory before doing this in case it screws up.

**Comment by icoulter** [comment]
Thanks. Looks like the squash worked.

**Comment by tlatorre-uchicago** [comment]
merged.

**Pull request closed without merging**